kube-runtime/src/finalizer.rs

```diff
@@ -207,3 +208,50 @@ pub enum Event<K> {
     /// - The grinch's heart grows a size or two
     Cleanup(K),
 }
+
+#[derive(Debug, Snafu)]
+pub enum DeleteError {
+    #[snafu(display("failed to delete object: {}", source))]
+    DeleteFailed { source: kube::Error },
+    #[snafu(display("failed to probe for whether the object is deleted yet: {}", source))]
+    DeleteProbeFailed { source: watcher::Error },
+}
+
+/// Try to delete an object, and wait for it to disappear from the Kubernetes API.
+/// If you do not wish to wait for the object to be deleted then you should use
+/// [`Api::delete`] instead.
+///
+/// This means that this function will not complete until all finalizers are done.
+///
+/// # Errors
+///
+/// This function fails if the deletion or status probe failed. In particular, it requires
+/// permission to use the following Kubernetes verbs:
+///
+/// - delete
+/// - list
+/// - watch
+///
+/// Note that the deletion will continue in the background if the probe fails.
+pub async fn finalize_and_delete<K>(api: &Api<K>, name: &str, dp: &DeleteParams) -> Result<(), DeleteError>
+where
+    K: Clone + DeserializeOwned + Debug + Send + Resource + 'static,
+{
+    api.delete(name, dp).await.context(DeleteFailed)?;
+    let watcher = watcher(api.clone(), ListParams {
+        field_selector: Some(format!("metadata.name={}", name)),
+        ..Default::default()
+    });
```

Comment:
This looks like it's combining concepts from [`kubectl delete`](https://github.com/kubernetes/kubectl/blob/master/pkg/cmd/delete/delete.go).

The `api.delete` invocation and stuff from the [cmd/wait module](https://pkg.go.dev/k8s.io/kubectl@v0.22.1/pkg/cmd/wait).
I think this is a good helper, but the combination of wait + api use here feels a bit awkward, and not very extensible.

I feel it would probably would be better to have a `wait` util mod in `kube_runtime` that lets you wait for conditions in a more generic manner. This way we could have `kube::utils` wrap the two long term, but have `kube_runtime` deal with these the watching. It would mean a little more sketching out of things first though.

Reply:
Yeah, I was thinking of breaking out the wait-for-deletion as something usable standalone function (at least internally at first).

As for `wait`, it /feels/ like the more "Rusty" way here would be some `Stream` combinator chain (though I don't have quite the right incantation right now...). I'm not sure we could add that much over that?

Reply:
Then again, it /would/ let us abstract out the desyncing ("is this the `Option<K>` you're looking for?")...

Reply:
Purely as a user of the library I believe it'd be good to have a method by this (or a similar) name.
That doesn't mean we can improve the implementation later on though.

Or do you believe we shouldn't even expose a method like this and only the building blocks individually?

Reply:
Definitely want to have a method like this, but I think the building blocks should be available. Particularly since there's equivalents in `kubectl`. I definitely think that a way to wait for deletion is usable for much more than just finalizers.

Anyway, I don't want to block this, because it's clearly useful, but I also think that we can probably find a nicer interface without too much work. Have raised #632 for sketching this out a little better.